### PR TITLE
RHIROS #188 Migrate to Clowder

### PR DESCRIPTION
**Pull request by apuntamb**

### CODE REVIEW

File: deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,109 @@
+---
```
Comment (kgaikwad) [inline deployment.yaml:1]: @apuntamb, what will be the expected value for `ENABLE_RBAC` in this case?

Comment (apuntamb) [reply deployment.yaml:1]: I will define the `ENABLE_RBAC` flag in the parameters with value `True`.

File: clowdapp.yaml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            cpu: ${CPU_REQUEST}
+            memory: ${MEMORY_REQUEST}
+          limits:
+            cpu: ${CPU_LIMIT}
+            memory: ${MEMORY_LIMIT}
+        env:
+          - name: CLOWDER_ENABLED
+            value: ${CLOWDER_ENABLED}
+    database:
+      name: ros-db
```
Comment (BlakeHolifield) [inline clowdapp.yaml:69]: This can stay `ros` if you'd like, it will have "db" appended to it by Clowder.

File: clowdapp.yaml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  spec:
+    envName: ${ENV_NAME}
+    dependencies:
+    - rbac
+    pods:
+      - name: backend
+        minReplicas: 1
+        image: ${IMAGE}:${IMAGE_TAG}
+        command: ["bash"]
+        args: ["-c", "python -m manage db upgrade && python -m ros.api.main"]
```
Comment (BlakeHolifield) [inline clowdapp.yaml:20]: I'm assuming this is a Django app? If so I've seen other apps break out their db migrations into init containers. If your app is setup to always handle `db upgrade`, then this is fine. Just a thought.

File: clowdapp.yaml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            cpu: ${CPU_REQUEST}
+            memory: ${MEMORY_REQUEST}
+          limits:
+            cpu: ${CPU_LIMIT}
+            memory: ${MEMORY_LIMIT}
+        livenessProbe:
+          failureThreshold: 3
+          httpGet:
+            path: /api/ros/v0/status
+            port: 8080
```
Comment (BlakeHolifield) [inline clowdapp.yaml:32]: One thing to note is that your app needs to run on port `8000` now and your metrics will need to be exposed on `9000`. 

Docs for that decision are here: https://clouddot.pages.redhat.com/docs/dev/getting-started/migration/faq.html#_why_does_my_app_have_to_run_on_port_8000 

So you'll just need to update: https://github.com/RedHatInsights/ros-backend/blob/e056b7ac9b12bc334f185ad420b7c9bda4610f57/ros/api/main.py#L11 to be `8000`.

**Review by BlakeHolifield — CHANGES_REQUESTED** [review]
Overall this looks great! You'll need to change the backend to serve on port `8000` instead of `8080`, but that's the only issue. Once that's done you should be good to go.

File: clowdapp.yaml [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+          - name: CLOWDER_ENABLED
+            value: ${CLOWDER_ENABLED}
+          - name: ENABLE_RBAC
+            value: "${ENABLE_RBAC}"
+    - name: processor
+      podSpec:
+        image: ${IMAGE}:${IMAGE_TAG}
+        minReplicas: ${{MIN_REPLICAS}}
+        command: ["bash"]
+        args: ["-c", "python -m ros.processor.main"]
```
Comment (apuntamb) [inline clowdapp.yaml:64]: Hi @BlakeHolifield! Do we need `initContainers` here as well?

File: clowdapp.yaml [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  required: true
+  value: 1Gi
+- description: Maximum amount of CPU the build container can use.
+  displayName: CPU Limit
+  name: CPU_LIMIT
+  required: true
+  value: '1'
+- description: Is clowder enabled
+  name: CLOWDER_ENABLED
+  value: "False"
```
Comment (apuntamb) [inline clowdapp.yaml:123]: @BlakeHolifield Should this be set to `true` here?

**Comment by apuntamb** [comment]
@BlakeHolifield as discussed over call, once all the queries are resolve, should we merge this PR?

File: clowdapp.yaml (hunk 6 above)
Comment (BlakeHolifield) [reply clowdapp.yaml:123]: I'd leave it as "false" for now until you get everything migrated. We don't want to misconfigure the currently running instances.

File: ros/api/main.py [hunk 7]
```diff
@@ -8,4 +8,4 @@
 initialize_routes(api)
 
 if __name__ == "__main__":
-    app.run(host='0.0.0.0', port=8080)
+    app.run(host='0.0.0.0', port=8000)
```
Comment (BlakeHolifield) [inline ros/api/main.py:11]: Do you have any apps that depend on your service? Just making sure before you change the port here.

File: clowdapp.yaml (hunk 5 above)
Comment (apuntamb) [reply clowdapp.yaml:64]: okay!

File: clowdapp.yaml (hunk 6 above)
Comment (apuntamb) [reply clowdapp.yaml:123]: gotcha, thanks!

File: ros/api/main.py (hunk 7 above)
Comment (apuntamb) [reply ros/api/main.py:11]: Yeah so we have all these services running for us: https://github.com/RedHatInsights/ros-backend/blob/main/scripts/docker-compose.yml

**Comment by apuntamb** [comment]
@kgaikwad Could you do a high level review of the PR, especially with respect to the port changes too?

File: scripts/docker-compose.yml [hunk 8]
```diff
@@ -119,12 +119,12 @@ services:
     depends_on:
       - insights-inventory-mq
     ports:
-      - 8081:8080
+      - 8001:8000
```
Comment (kgaikwad) [inline scripts/docker-compose.yml:122]: For ROS and here, you mentioned port value  `8000`, will it create any problem?

**Review by kgaikwad — CHANGES_REQUESTED** [review]
@apuntamb, added one inline comment about port value.
I didn't test the changes.

File: scripts/docker-compose.yml (hunk 8 above)
Comment (apuntamb) [reply scripts/docker-compose.yml:122]: @kgaikwad I guess it slipped my mind that we have already modified the ports for concerned services in the file.
Could you try running the containers and ros api/processors once with this change, to see if it works for you well?

Comment (kgaikwad) [reply scripts/docker-compose.yml:122]: Sure, I will verify it once. Is dev setup working fine on your local with mentioned ports change?

File: Makefile [hunk 9]
```diff
@@ -8,18 +8,11 @@ metadata={"branch_info": {"remote_branch": -1, "remote_leaf": -1}, "bios_uuid":
 identity={"identity": {"account_number": "0000001","type": "User","user": {"username": "tuser@redhat.com","email": "tuser@redhat.com","first_name": "test","last_name": "user","is_active": true,"is_org_admin": false, "is_internal": true, "locale": "en_US"},"internal": { "org_id": "000001"}}}
 b64_identity=$(shell echo '${identity}' | base64 -w 0 -)
 
-
-ros-upload-data:
-	curl -vvvv -F "upload=@sample-files/ros-sample.tar.gz;type=application/vnd.redhat.resource-optimization.sample+tgz" -F 'metadata=${metadata}' \
-		-H "x-rh-identity: ${b64_identity}" \
-		-H "x-rh-request_id: testtesttest" \
-		localhost:3000/api/ingress/v1/upload  | python -m json.tool
-
```
Comment (apuntamb) [inline Makefile:17]: I hope we no longer need this command, right @kgaikwad ?

File: scripts/docker-compose.yml (hunk 8 above)
Comment (apuntamb) [reply scripts/docker-compose.yml:122]: Yeah its working in my local with the ports change :)

**Review by kgaikwad — APPROVED** [review]
Tested docker-compose file change. 
I can see all required services up and running. 
Tested archive upload and found data stored correctly into DB.
No further comments from my end. 

@BlakeHolifield - do you have any additional comments?

File: clowdapp.yaml [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    deployments:
+    - name: backend
+      podSpec:
+        image: ${IMAGE}:${IMAGE_TAG}
+        minReplicas: 1
+        initContainers:
+          - command:
+            - sh
+            - -c
+              python -m manage db upgrade && python -m ros.api.main
```
Comment (BlakeHolifield) [inline clowdapp.yaml:24]: I think you need an extra `-` here.

Comment (apuntamb) [reply clowdapp.yaml:24]: But, isn't it something like this - https://github.com/RedHatInsights/vulnerability-engine/blob/master/deploy/clowdapp.yaml#L230-L235

Comment (BlakeHolifield) [reply clowdapp.yaml:24]: Right. You can do it that way as well, but I don't think the current version works without that `- >-` line before the python code.

File: clowdapp.yaml [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    deployments:
+    - name: backend
+      podSpec:
+        image: ${IMAGE}:${IMAGE_TAG}
+        minReplicas: 1
+        initContainers:
+          - command:
+            - sh
+            - -c
+            - python -m manage db upgrade && python -m ros.api.main
```
Comment (apuntamb) [inline clowdapp.yaml:24]: @BlakeHolifield Is this the correct syntax?

**Comment by kgaikwad** [comment]
From the offline discussion with @apuntamb, this pull-request is ready to go. So merging it.
Thank you @BlakeHolifield for review and @apuntamb  for addressing all comments.

**Pull request merged**